src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSObject.cs

```diff
@@ -76,6 +76,80 @@ public object Invoke(string method, params object?[] args)
             return res;
         }
 
+        public struct EventListenerOptions {
+            public bool Capture;
+            public bool Once;
+            public bool Passive;
+            public bool Weak;
+            public object? Signal;
+        }
+
+        // FIXME: The existing Runtime binding apis are completely unusable for this
+        private static int _NextJSObjectID = 0xFFFFFF;
+        private static JSObject GetJSObjectForDelegate (Delegate d) {
+            var result = Runtime.TryGetJSObjectForCLRObject(d);
+            if (result != null)
+                return result;
+
+            int id = _NextJSObjectID++;
+            return Runtime.CreateJSObjectForCLRObject(d, id);
+        }
+
+        public JSObject AddEventListener(string name, Delegate listener, EventListenerOptions? options = null)
+        {
+            var optionsDict = options.HasValue
+                ? new JSObject()
+                : null;
+
+            try {
+                if (options?.Signal != null)
+                    throw new NotImplementedException("EventListenerOptions.Signal");
+
+                var jsfunc = GetJSObjectForDelegate(listener);
+                int exception;
+                if (options.HasValue) {
+                    // TODO: Optimize this
+                    var _options = options.Value;
+                    optionsDict?.SetObjectProperty("capture", _options.Capture, true, true);
+                    optionsDict?.SetObjectProperty("once", _options.Once, true, true);
+                    optionsDict?.SetObjectProperty("passive", _options.Passive, true, true);
+                }
+
+                // TODO: Unique optimized syscall
+                // Interop.Runtime.AddEventListener(JSHandle, name, listener, optionsDict);
+                var args = options.HasValue
+                    ? new object?[] { jsfunc, optionsDict }
+                    : new object?[] { jsfunc };
+                var res = Interop.Runtime.InvokeJSWithArgs(JSHandle, "addEventListener", args, out exception);
+                if (exception != 0)
+                    throw new JSException((string)res);
+                else if (options?.Weak != true) {
+                    bool success = false;
+                    jsfunc.DangerousAddRef(ref success);
```

Comment:
I think it could not be really weak as the JS doesn't know if C# object want to be collected.

Reply:
The weak info will also be passed through to JS once i introduce the icall

Reply:
If weak is true and C# side get's collected first we are in trouble.